### PR TITLE
Mobile client FFI

### PR DESCRIPTION
**Pull request by Robert-Steiner**
Summary:

* added the new mobile client FFI 
* removed the old FFI 
* added C-tests

### CODE REVIEW

File: rust/xaynet-ffi/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+- run `RUSTUP_TOOLCHAIN=nightly cbindgen --output xaynet_ffi.h`
+
+
+## Run tests
+
+### macOS
+
+- run `cargo build`
+- run `cc -o tests/ffi_test.o -Wl,-dead_strip -I. tests/ffi_test.c ../target/debug/libxaynet_sdk.a -framework Security -framework Foundation`
```
Comment (little-dude) [inline rust/xaynet-ffi/README.md:17]: what are the `-framework Security -framework Foundation` arguments?

File: rust/xaynet-ffi/src/lib.rs [hunk 2]
```diff
@@ -1,812 +1,710 @@
...
 //! # Safety
+//!
 //! Many functions of this module are marked as `unsafe` to explicitly announce the possible
 //! unsafety of the function body as well as the return value to the caller. At the same time,
 //! each `unsafe fn` uses `unsafe` blocks to precisely pinpoint the sources of unsafety for
 //! reviewers (redundancy warnings will be fixed by [#69173]).
 //!
-//! **Note, that the `unsafe` code has not been externally audited yet!**
+//! Most of the functions have one / or more pointers as parameters.
+//! All functions ensure null-safety for these pointers.
```
Comment (little-dude) [inline rust/xaynet-ffi/src/lib.rs:11]: Does "null-safety" refer to a well defined Rust or `ffi-support` concept, or does it just mean that we check whether the pointer is null and return an error if it is?

File: rust/xaynet-ffi/src/lib.rs [hunk 3]
```diff
@@ -1,812 +1,710 @@
...
+//! However, when calling this function, you need to ensure that a pointer:
+//! - is properly aligned,
```
Comment (little-dude) [inline rust/xaynet-ffi/src/lib.rs:13]: What does _properly aligned_ mean for us? I'm not familiar with this so I've skimmed [Wikipedia's data structure alignment](https://en.wikipedia.org/wiki/Data_structure_alignment) page and I don't fully understand the alignment restrictions that the pointer must conform to.

File: rust/xaynet-ffi/src/lib.rs [hunk 4]
```diff
@@ -1,812 +1,710 @@
...
-pub struct PrimitiveModel {
-    /// A raw mutable pointer to an array of primitive values.
-    pub ptr: *mut c_void,
-    /// The length of that array.
-    pub len: c_ulonglong,
-    /// The data type of the array's elements.
-    pub dtype: c_uint,
-}
+/// A Opaque type of MobileClient.
+/// see [FFI-C-OPAQUE](https://anssi-fr.github.io/rust-guide/07_ffi.html#recommendation-a-idffi-c-opaqueaffi-c-opaque)
```
Comment (little-dude) [inline rust/xaynet-ffi/src/lib.rs:61]: It is very nice to have the reference but might be confusing for people reading the doc. Maybe we could turn it into a comment.

File: rust/xaynet-ffi/README.md (hunk 1 above)
Comment (finiteprods) [reply rust/xaynet-ffi/README.md:17]: when I try to run this:
```
cc: error: Security: No such file or directory
cc: error: Foundation: No such file or directory
cc: error: unrecognized command line option ‘-framework’
cc: error: unrecognized command line option ‘-framework’
```
I'm not on macOS. Is this macOS-specific?

Comment (Robert-Steiner) [reply rust/xaynet-ffi/README.md:17]: that is right, it is macOS-specific

**Review by little-dude — APPROVED** [review]

**Pull request merged**